libmamba/include/mamba/core/graph_util.hpp

```diff
@@ -7,303 +7,493 @@
 
         template <class V>
-        void depth_first_search_impl(V& visitor, node_id node, color_list& colors) const;
+        void depth_first_search_impl(V& visitor,
+                                     node_id node,
+                                     visited_list& status,
+                                     adjacency_list const& successors) const;
 
         node_list m_node_list;
-        adjacency_list m_adjacency_list;
+        adjacency_list m_predecessors;
+        adjacency_list m_successors;
     };
 
     template <class G>
     class default_visitor
     {
     public:
-        using graph_type = G;
-        using node_id = typename graph_type::node_id;
+        using graph_t = G;
+        using node_id = typename graph_t::node_id;
 
-        void start_node(node_id, const G&)
+        void start_node(node_id, const graph_t&)
         {
         }
-        void finish_node(node_id, const G&)
+        void finish_node(node_id, const graph_t&)
         {
         }
 
-        void start_edge(node_id, node_id, const G&)
+        void start_edge(node_id, node_id, const graph_t&)
         {
         }
-        void tree_edge(node_id, node_id, const G&)
+        void tree_edge(node_id, node_id, const graph_t&)
         {
         }
-        void back_edge(node_id, node_id, const G&)
+        void back_edge(node_id, node_id, const graph_t&)
         {
         }
-        void forward_or_cross_edge(node_id, node_id, const G&)
+        void forward_or_cross_edge(node_id, node_id, const graph_t&)
         {
         }
-        void finish_edge(node_id, node_id, const G&)
+        void finish_edge(node_id, node_id, const graph_t&)
         {
         }
     };
 
-    template <class G>
-    class predecessor_recorder : private default_visitor<G>
+    template <typename Node, typename Edge = void>
+    class DiGraph : public DiGraphBase<Node, DiGraph<Node, Edge>>
     {
     public:
-        using base_type = default_visitor<G>;
-        using graph_type = typename base_type::graph_type;
-        using node_id = typename base_type::node_it;
-        using predecessor_map = std::map<node_id, node_id>;
+        using Base = DiGraphBase<Node, DiGraph<Node, Edge>>;
+        using edge_t = Edge;
+        using node_id = typename Base::node_id;
+        using edge_id = std::pair<node_id, node_id>;
+        using edge_map = std::map<edge_id, edge_t>;
 
-        using base_type::back_edge;
-        using base_type::finish_node;
-        using base_type::forward_or_cross_edge;
-        using base_type::start_edge;
-        using base_type::start_node;
+        void add_edge(node_id from, node_id to, edge_t const& data);
+        void add_edge(node_id from, node_id to, edge_t&& data);
 
-        void tree_edge(node_id from, node_id to, const G&);
-        void finish_edge(node_id id, const G&);
-
-        const predecessor_map& get_predecessors() const;
+        const edge_map& edges() const;
+        edge_map& edges();
```

Comment:
Switching for
```cpp
        node_t const& node(node_id id) const;
        node_t & node(node_id id);
```
Since changing single node data is allowed and needed.